findspam.py

```diff
@@ -617,6 +627,30 @@ def mostly_img(s, site):
     return False, ""
 
 
+@create_rule("post is likely nonsense", title=False,
+             sites=["codegolf.stackexchange.com",
+                    "stackoverflow.com", "ja.stackoverflow.com", "pt.stackoverflow.com",
+                    "es.stackoverflow.com", "islam.stackexchange.com",
+                    "japanese.stackexchange.com", "anime.stackexchange.com",
+                    "hinduism.stackexchange.com", "judaism.stackexchange.com",
+                    "buddhism.stackexchange.com", "chinese.stackexchange.com",
+                    "french.stackexchange.com", "spanish.stackexchange.com",
+                    "portuguese.stackexchange.com", "korean.stackexchange.com",
+                    "ukrainian.stackexchange.com", "italian.stackexchange.com"],
+             max_rep=10000, max_score=10000)
+def nonsense(s, site):
+    if len(s) == 0:
+        return False, ""
+    if "pytest" in sys.modules:
+        return False, ""
+    probability = [float(s.count(x)) / len(s) for x in s]
+    entropy_per_char = -sum([x * math.log2(x) for x in probability]) / len(s)
```

Comment:
I have had exactly one and a half weeks of formal statistics education in my life, so take what I say with a grain of salt -- but I believe there's a subtle mathematical error here that's leading to a lot of really weird false positives.

This definition of entropy appears to be the same one that Wikipedia uses, which is:

![](https://wikimedia.org/api/rest_v1/media/math/render/svg/bfe3616dee43f6287d4a4e2a557de8d48ad24926)

...where X is a [discrete random variable] -- in other words, a variable that can randomly take on elements of a set of discrete events. In this case, X is a discrete random variable representing the possible outcomes of randomly choosing a single character from a post. For example, if 25% of characters in a post are spaces, then X has a 25% chance of being a space.

According to this definition, in order to calculate the entropy of a discrete random variable, first you calculate the "suprisingness" of each possible event, where surprisingness is defined as -log(P(X)) (so that lower-probability events are more surprising). Then, the entropy is the "expected surprisingness" of the random variable, which is the average surprisingness of the events weighted by probability. That's equivalent to the average number of bits of information you learn on each sampling of the random variable -- or the average number of bits of information carried by each character in the input post, which is what we're after.

There are two errors in this implementation:

1. The equation used is for entropy for a single random variable, and our variable is a single character. Therefore, the result **is already entropy-per-character** and dividing by `len(s)` is incorrect. 

2. **`probability` includes duplicate values** whenever the post uses the same character more than once, while the sum in the definition of entropy is supposed to be across *distinct* events of the discrete random variable. By including duplicates, instead of the expected `P(x) log(P(X))`, we're calculating `n * P(x) log(P(x))` (where *n* is the number of occurrences of the character).

Since `P(x) = n / len(s)`, the value we're currently calculating is equal to `-sum(P(x)^2 len(s) log(P(x)) / len(s)`. The `len(s)`'s cancel out, so we end up with <code>-sum(P(x)<strong>**^2**</strong> log(P(x))</code>. Therefore, we can kill two birds with one stone and fix both errors by simply changing the line to: 

    entropy_per_char = -sum([math.log2(x) for x in probability]) / len(s)

With this change, the `entropy_for_char` for most posts ends up being around 4 bits. This value seems a lot more reasonable than 0.25 (a coin flip is 1 bit of entropy, and picking a random character from a post should *definitely* have more entropy than a coin flip). However, this means the `ENTROPY_TOO_HIGH/LOW` constants will have to be tweaked.

Reply:
Looks like I managed to forget that `for x in "string"` does not remove duplicate, and added the `len(s)` part when the result turned out to be erroneous.